src/Parser/Druid/Feral/Modules/FeralCore/Snapshot.js

```diff
@@ -0,0 +1,127 @@
+import SPELLS from 'common/SPELLS';
+import Analyzer from 'Parser/Core/Analyzer';
+import Combatants from 'Parser/Core/Modules/Combatants';
+import { encodeTargetString } from 'Parser/Core/Modules/EnemyInstances';
+
+const debug = false;
+
+/*
+Feral has a snapshotting mechanic which means the effect of some buffs are maintained over the duration of
+some DoTs even after the buff has worn off.
+Players should follow a number of rules with regards when they refresh a DoT and when they do not, depending
+on what buffs the DoT has snapshot and what buffs are currently active.
+
+RAKE snapshots:
+  Prowl, Tiger's Fury, Bloodtalons
+RIP snapshots:
+  Tiger's Fury, Bloodtalons
+MOONFIRE_FERAL snapshots:
+  Tiger's Fury
+THRASH_FERAL snapshots: (but isn't used in single target situations)
+  Tiger's Fury, Moment of Clarity
+
+The Snapshot class is 'abstract', and shouldn't be directly instantiated. Instead classes should extend
+it and test how well the combatant is making use of the snapshot mechanic.
+*/
+
+const TIGERS_FURY_MULTIPLIER = 1.15;
+const PROWL_MULTIPLIER = 2.00; // also applied by Incarnation: King of the Jungle, and Shadowmeld
+const BLOODTALONS_MULTIPLIER = 1.20;
+
+const JAGGED_WOUNDS_MODIFIER = 0.80;  // "[...]deal the same damage as normal but in 20% less time."
+const PANDEMIC_FRACTION = 0.3;
+
+// leeway in ms after loss of bloodtalons/prowl buff to count a cast as being buffed. Keep well below GCD to avoid false positives.
+const BUFF_WINDOW_TIME = 300;
+
+class Snapshot extends Analyzer {
+  static dependencies = {
+    combatants: Combatants,
+  };
+
+  // extending class should fill these in:
+  spellCastId;
+  isProwlAffected;
+  isTigersFuryAffected;
+  isBloodtalonsAffected;
+  durationOfFresh;
+
+  // initialised to before combat log starts to avoid false positives
+  bloodtalonsFadedAt = -10000;
+  prowlFadedAt = -10000;
+
+  stateByTarget = [];
+
+  // It's common for buffs that are consumed by an ability to be reported as not present when the ability's event is processed.
+  // So record when the buff wears off, and if that's sufficiently close to the ability being used we can assume it was active.
+  on_byPlayer_removebuff(event) {
+    const spellId = event.ability.guid;
+    if (this.isProwlAffected && (
+        SPELLS.PROWL.id === spellId ||
+        SPELLS.PROWL_INCARNATION.id === spellId || 
+        SPELLS.SHADOWMELD.id === spellId)) {
+      this.prowlFadedAt = event.timestamp;
+    } else if (this.isBloodtalonsAffected && SPELLS.BLOODTALONS_BUFF.id === spellId) {
+      this.bloodtalonsFadedAt = event.timestamp;
+    }
+  }
+
+  on_byPlayer_cast(event) {
```

Comment:
I recommend using `on_byPlayer_applydebuff` and `on_byPlayer_refreshdebuff` and using the debuffID instead of `on_byPlayer_cast` and castID. Cast might not always be accurate because the debuff might not be applied (immune, dodged, parry, etc.) or removed early (pretty rare in pve but might aswell cover that case). It would also support anything that allows you to apply the bleed to multiple targets like Thrash.

Reply:
For marking the cast as inefficient on the timeline you can just save the most recent cast at all times (The cast should always come before the debuff apply event).

Reply:
Thank you, the cast not actually applying a DoT is important to catch.

Using `on_byPlayer_applydebuff` and `on_byPlayer_refreshdebuff` instead of `on_byPlayer_cast` means it loses easy access to the cast event needed to mark inefficient casts on the timeline. So I made some further changes to link the cast event with the DoT's state information. Usually the debuff event occurs after the cast but in some logs it occurs before, so it attempts to make the link during both debuff and cast event handling.

As the "rules" should only be checked after we have the cast event, `checkRuleIfReady` has been added. This also makes sure each DoT application only gets checked once to avoid any double-counting issues. I packed all the information the rule check needs into the DoT's state object, making it easier to call the rule check whenever the cast event is added to it.